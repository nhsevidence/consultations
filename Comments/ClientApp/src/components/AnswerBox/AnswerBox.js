// @flow

import React, { Component, Fragment } from "react";

type PropsType = {
	staticContext?: any,
	isVisible: boolean,
	answer: AnswerType,
	readOnly: boolean,
	saveAnswerHandler: Function,
	deleteAnswerHandler: Function,
	unique: string,
	updateUnsavedIds: Function,
};

type StateType = {
	answer: AnswerType,
	unsavedChanges: boolean,
};

export class AnswerBox extends Component<PropsType, StateType> {
	constructor() {
		super();
		this.state = {
			answer: {},
			unsavedChanges: false,
		};
	}

	componentDidMount() {
		this.setState({
			answer: this.props.answer,
		});
	}

	textareaChangeHandler = (e: SyntheticEvent) => {
		const answer = this.state.answer;
		answer.answerText = e.target.value;
		const unsavedChanges = !(answer.answerId === -1 && answer.answerText.length === 0);
		this.props.updateUnsavedIds(`${answer.questionId}q`, unsavedChanges);
		this.setState({
			answer,
			unsavedChanges,
		});
	};

	static getDerivedStateFromProps(nextProps: any, prevState: any) {
		const prevTimestamp = prevState.answer.lastModifiedDate;
		const nextTimestamp = nextProps.answer.lastModifiedDate;
		const hasAnswerBeenUpdated = () => prevTimestamp !== nextTimestamp;
		if (hasAnswerBeenUpdated()) {
			return {
				answer: nextProps.answer,
				unsavedChanges: false,
			};
		}
		return null;
	}

	render() {
		if (!this.state.answer) return null;
		const {
			answerText,
			answerId,
			questionId,
		} = this.state.answer;
		const unsavedChanges = this.state.unsavedChanges;
		const answer = this.state.answer;
		const readOnly = this.props.readOnly;

		return (

			<Fragment>
				<section role="form">
					<form onSubmit={e => this.props.saveAnswerHandler(e, answer, this.props.questionId)} className="mb--0">
						<div className="form__group form__group--textarea mb--b">
							<textarea
								data-qa-sel="Comment-text-area"
								disabled={readOnly}
								id={this.props.unique}
								className="form__input form__input--textarea"
								onInput={this.textareaChangeHandler}
								defaultValue={answerText}/>
						</div>
						{!readOnly && answerText && answerText.length > 0 ?
							unsavedChanges ?
								<input
									data-qa-sel="submit-button"
									className="btn ml--0 mb--0"
									type="submit"
									value="Save answer"
								/>
								:
								<span data-qa-sel="comment-saved" className="ml--0 mb--0 CommentBox__savedIndicator">Saved</span>
							:
							null
						}
						{!readOnly && answerId > 0 &&
						<button
							data-qa-sel="delete-comment-button"
							className="btn mr--0 mb--0 right"
							onClick={e => this.props.deleteAnswerHandler(e, questionId, answerId)}>
							Delete
						</button>
						}
					</form>
				</section>
			</Fragment>
		);
	}
}
