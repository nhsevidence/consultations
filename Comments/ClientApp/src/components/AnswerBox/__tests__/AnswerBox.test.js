/* global jest */
import React from "react";
import { shallow, mount } from "enzyme";
import { AnswerBox } from "../AnswerBox";
import answerWithAnswer from "./answerWithAnswer.json";
import answerWithoutAnswer from "./answerWithoutAnswer.json";
import toJson from "enzyme-to-json";

describe("[ClientApp] ", () => {
	describe("AnswerBox Component", () => {

		const answerPropsWithAnswer = {
			isVisible: true,
			answer: answerWithAnswer,
			readOnly: false,
			saveHandler: jest.fn(),
			deleteHandler: jest.fn(),
			unique: "string",
		};

		const answerPropsWithoutAnswer = {
			isVisible: true,
			answer: answerWithoutAnswer,
			readOnly: false,
			saveHandler: jest.fn(),
			deleteHandler: jest.fn(),
			unique: "string",
		};

		it("sets text area with comment text correctly", () => {
			const wrapper = shallow(<AnswerBox {...answerPropsWithAnswer} />);
			expect(wrapper.find("textarea").length).toEqual(1);
			expect(wrapper.find("textarea").props().value).toEqual("some answer text");
		});

		it("should update UnsavedChanges if lastupdateddate has changed", () => {
			const wrapper = mount(<AnswerBox {...answerPropsWithAnswer} />);
			wrapper.setState({unsavedChanges: true});
			const updatedProps = {
				answer: {
					answerId: answerWithoutAnswer.answerId,
					answerText: "an updated answer",
					lastModifiedDate: new Date("02/04/2018").toISOString(),
				},
			};
			wrapper.setProps(updatedProps);
			expect(wrapper.state().unsavedChanges).toEqual(false);
		});

		it("should not update UnsavedChanges if lastupdateddate has not changed", () => {
			const wrapper = mount(<AnswerBox {...answerPropsWithAnswer} />);
			wrapper.setState({unsavedChanges: true});
			wrapper.setProps(answerPropsWithAnswer);
			expect(wrapper.state().unsavedChanges).toEqual(true);
		});

		it("updated comment text in state after new props received", () => {
			const wrapper = mount(<AnswerBox {...answerPropsWithAnswer} />);
			const updatedProps = {
				answer: {
					answerId: answerWithoutAnswer.answerId,
					answerText: "an updated answer",
				},
			};
			wrapper.setProps(updatedProps);
			expect(wrapper.state().answer.answerText).toEqual("an updated answer");
		});

		it("should match snapshot with answer", () => {
			const wrapper = mount(
				<AnswerBox {...answerPropsWithAnswer} />
			);
			expect(toJson(wrapper, {
				noKey: true,
				mode: "deep",
			})).toMatchSnapshot();			
		});

		it("should match snapshot without answer", () => {
			const wrapper = mount(
				<AnswerBox {...answerPropsWithoutAnswer} />
			);
			expect(toJson(wrapper, {
				noKey: true,
				mode: "deep",
			})).toMatchSnapshot();			
		});
	});
});
