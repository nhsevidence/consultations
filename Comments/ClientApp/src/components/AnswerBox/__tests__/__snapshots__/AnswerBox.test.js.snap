// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  AnswerBox Component should match snapshot with answer 1`] = `
<section
  role="form"
>
  <form
    className="mb--0"
    onSubmit={[Function]}
  >
    <div
      className="form__group form__group--textarea mb--b"
    >
      <textarea
        className="form__input form__input--textarea"
        data-hj-whitelist={true}
        data-qa-sel="Comment-text-area"
        defaultValue="an updated answer"
        disabled={false}
        id="string"
        onInput={[Function]}
      />
    </div>
    <span
      className="ml--0 mb--0 CommentBox__savedIndicator"
    >
      Saved
    </span>
    <button
      className="btn mr--0 mb--0 right"
      data-qa-sel="delete-comment-button"
      onClick={[Function]}
    >
      Delete
    </button>
  </form>
</section>
`;

exports[`[ClientApp]  AnswerBox Component should match snapshot without answer 1`] = `
<section
  role="form"
>
  <form
    className="mb--0"
    onSubmit={[Function]}
  >
    <div
      className="form__group form__group--textarea mb--b"
    >
      <textarea
        className="form__input form__input--textarea"
        data-hj-whitelist={true}
        data-qa-sel="Comment-text-area"
        defaultValue=""
        disabled={false}
        id="string"
        onInput={[Function]}
      />
    </div>
  </form>
</section>
`;
